discord/commands/mixins.py

```diff
@@ -0,0 +1,895 @@
+from __future__ import annotations
+
+import asyncio
+import datetime
+import functools
+from typing import (
+    TYPE_CHECKING,
+    TypeVar,
+    Callable,
+    Coroutine,
+    Optional,
+    Union,
+    Generic,
+    Any,
+    Dict,
+    List,
+)
+
+from .. import utils, abc
+from ..errors import (
+    ApplicationCommandError,
+    CheckFailure,
+    CommandError,
+    CommandInvokeError,
+    DisabledCommand,
+    CommandOnCooldown,
+)
+from .cooldowns import BucketType, CooldownMapping, MaxConcurrency
+
+if TYPE_CHECKING:
+    from typing_extensions import ParamSpec
+
+    from ..bot import Bot, AutoShardedBot
+    from ..user import User, ClientUser
+    from ..member import Member
+    from ..guild import Guild
+    from ..message import Message
+    from ..interactions import Interaction
+    from ..abc import MessageableChannel
+    from ..voice_client import VoiceProtocol
+    from ..state import ConnectionState
+    from ..cog import Cog
+
+    P = ParamSpec("P")
+else:
+    P = TypeVar("P")
+
+BotT = TypeVar("BotT", bound="Union[Bot, AutoShardedBot]")
+CogT = TypeVar("CogT", bound="Cog")
+CallbackT = TypeVar("CallbackT")
+ContextT = TypeVar("ContextT", bound="BaseContext")
+
+T = TypeVar("T")
+Coro = Coroutine[Any, Any, T]
+MaybeCoro = Union[T, Coro[T]]
+
+Check = Union[
+    Callable[[CogT, ContextT], MaybeCoro[bool]],
+    Callable[[ContextT], MaybeCoro[bool]],
+]
+
+Error = Union[
+    Callable[[CogT, "BaseContext[Any]", CommandError], Coro[Any]],
+    Callable[["BaseContext[Any]", CommandError], Coro[Any]],
+]
+ErrorT = TypeVar("ErrorT", bound="Error")
+
+Hook = Union[
+    Callable[[CogT, ContextT], Coro[Any]],
+    Callable[[ContextT], Coro[Any]]
+]
+HookT = TypeVar("HookT", bound="Hook")
+
+
```

Comment:
To stay consistent with the other files, `__all__` should be set in this file.

Reply:
Why was this resolved without any explanation?

Reply:
`__all__` was added

Reply:
Oh I didn't see, my bad